File: Dec2020/26Dec/SubdomainVisitCount.java
/*
Problem Name: 811. Subdomain Visit Count
Problem Link: https://leetcode.com/problems/subdomain-visit-count/
*/
class Solution {
    public List<String> subdomainVisits(String[] cpdomains) {
        int n = cpdomains.length;
        HashMap<String,Integer> map = new HashMap<String,Integer>();
        for(int i=0;i<n;i++){
            String inp[] = cpdomains[i].split(" ");
            int cnt = Integer.parseInt(inp[0]);
            String domains[] = inp[1].split("\\.");
            String dom = "";
            for(int j=domains.length-1;j>=0;j--){
                dom = "." + domains[j] + dom;
                map.put(dom.substring(1), map.getOrDefault(dom.substring(1), 0) + cnt);
            }
        }
        List<String> ans = new ArrayList<String>();
        for(String key : map.keySet()){
            ans.add(map.get(key)+" "+key);
        }
        return ans;
    }
}

File: Dec2020/26Dec/SubdomainVisitCount.cpp
/*
Problem name: 811. Subdomain Visit Count
Problem link: https://leetcode.com/problems/subdomain-visit-count/
*/

class Solution {
public:
    vector<string> subdomainVisits(vector<string>& cpdomains) {
        map<string, int> map;
        for (auto cd : cpdomains) {
            int space = cd.find(" ");
            int num = stoi(cd.substr (0, space));
            string s = cd.substr(space + 1);
            for (int i=0;i<s.size();++i)
                if (s[i] == '.')
                    map[s.substr(i+1)] += num;
            map[s] += num;
        }
        vector<string> res;
        for (auto k : map)
            res.push_back (to_string(k.second) + " " + k.first);
        return res;
    }
};
